lib/internal/test_runner/mock/mock_timers.js

```diff
@@ -0,0 +1,343 @@
+  ArrayPrototypeAt,
+  ArrayPrototypeForEach,
+  ArrayPrototypeIncludes,
+  DateNow,
+  FunctionPrototypeApply,
+  FunctionPrototypeBind,
+  Promise,
+  SymbolAsyncIterator,
+  globalThis,
+} = primordials;
+const {
+  validateAbortSignal,
+  validateArray,
+} = require('internal/validators');
+
+const {
+  AbortError,
+  codes: {
+    ERR_INVALID_STATE,
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+
+const PriorityQueue = require('internal/priority_queue');
+const nodeTimers = require('timers');
+const nodeTimersPromises = require('timers/promises');
+const EventEmitter = require('events');
+
+function compareTimersLists(a, b) {
+  return (a.runAt - b.runAt) || (a.id - b.id);
+}
+
+function setPosition(node, pos) {
+  node.priorityQueuePosition = pos;
+}
+
+function abortIt(signal) {
+  return new AbortError(undefined, { cause: signal.reason });
+}
+
+const SUPPORTED_TIMERS = ['setTimeout', 'setInterval'];
+
+class MockTimers {
+  #realSetTimeout;
+  #realClearTimeout;
+  #realSetInterval;
+  #realClearInterval;
+
+  #realPromisifiedSetTimeout;
+  #realPromisifiedSetInterval;
+
+  #realTimersSetTimeout;
+  #realTimersClearTimeout;
+  #realTimersSetInterval;
+  #realTimersClearInterval;
+
+  #timersInContext = [];
+  #isEnabled = false;
+  #currentTimer = 1;
+  #now = DateNow();
+
+  #executionQueue = new PriorityQueue(compareTimersLists, setPosition);
+
+  #setTimeout = FunctionPrototypeBind(this.#createTimer, this, false);
+  #clearTimeout = FunctionPrototypeBind(this.#clearTimer, this);
+  #setInterval = FunctionPrototypeBind(this.#createTimer, this, true);
+  #clearInterval = FunctionPrototypeBind(this.#clearTimer, this);
+
+  constructor() {
+    emitExperimentalWarning('The MockTimers API');
+  }
+
+  #createTimer(isInterval, callback, delay, ...args) {
+    const timerId = this.#currentTimer++;
+    this.#executionQueue.insert({
+      __proto__: null,
+      id: timerId,
+      callback,
+      runAt: DateNow() + delay,
+      interval: isInterval,
+      args,
+    });
+
+    return timerId;
+  }
+
+  #clearTimer(position) {
+    this.#executionQueue.removeAt(position);
+  }
+
+  async * #setIntervalPromisified(interval, startTime, options) {
+    const context = this;
+    const emitter = new EventEmitter();
+    if (options?.signal) {
+      validateAbortSignal(options.signal, 'options.signal');
+
+      if (options.signal?.aborted) {
+        throw abortIt(options.signal);
+      }
+
+      const onabort = (reason) => {
+        emitter.emit('data', { aborted: true, reason });
+        options?.signal?.removeEventListener('abort', onabort);
+      };
+
+      options.signal?.addEventListener('abort', onabort);
+    }
+
+    const eventIt = EventEmitter.on(emitter, 'data');
+    const callback = () => {
+      startTime += interval;
+      emitter.emit('data', startTime);
+    };
+
+    const timerId = this.#createTimer(true, callback, interval, options);
+
+    const iterator = {
```

Comment:
can this be simplified?
```
const { on } = require('node:events');

const events = on(emitter, 'data', { signal });
for await (const result of events) {
	yield result.value
}

Reply:
It's probably fine to have this as a for await and just `.call it with a safe async iterator if primordials are a concern here (for me they are not).

Moreover, I don't think the reasoning for primordials (to have consistent internal behavior) is relevant for a mocking API anyway?